File: cloud-run/KUCOIN/asia-kucoin-exec-handler/exchanges/shuffle.py
from decimal import Decimal, getcontext

async def rearrange_tps(quantity, precision, tps_amount, min_qty):
    
    getcontext().prec = precision + 5  # set precision to avoid rounding errors

    rounded_quantity = Decimal(str(quantity))
    min_qty = Decimal(str(min_qty))

    tps_amount = [Decimal(str(tp)) for tp in tps_amount]
    tps_sum = sum(tps_amount)

    tps_ratio = [tp / tps_sum for tp in tps_amount]
    tps_rounded = [(rounded_quantity * ratio).quantize(Decimal('1e-{0}'.format(precision))) for ratio in tps_ratio]

    # Adjust values if any tps_rounded value is smaller than min_qty
    for i in range(len(tps_rounded)):
        if tps_rounded[i] < min_qty:
            tps_rounded[i] = min_qty

    tps_sum_rounded = sum(tps_rounded)

    # Adjust total quantity if necessary
    if tps_sum_rounded > rounded_quantity:
        excess = tps_sum_rounded - rounded_quantity

        # Find the indices of the last values in tps_rounded
        last_indices = [i for i in range(len(tps_rounded) - 1, -1, -1)]

        # Iterate over the last indices in reverse order
        for index in last_indices:
            if tps_rounded[index] >= excess:
                tps_rounded[index] -= excess
                break
            else:
                excess -= tps_rounded[index]
                tps_rounded[index] = Decimal('0')

    elif tps_sum_rounded < rounded_quantity:
        deficit = rounded_quantity - tps_sum_rounded

        # Find the indices of the last values in tps_rounded
        last_indices = [i for i in range(len(tps_rounded) - 1, -1, -1)]

        # Iterate over the last indices in reverse order
        for index in last_indices:
            if tps_rounded[index] > 0:
                tps_rounded[index] += deficit
                break

    return [float(val) for val in tps_rounded]

File: cloud-functions/submitTrade-js/utils/shuffle.js
const { Decimal } = require('decimal.js');

async function rearrangeTps(quantity, precision, tpAmounts, minQty) {
  
  Decimal.set({ precision: precision + 5 }); // Set precision to avoid rounding errors

  const roundedQuantity = new Decimal(quantity);
  const minQuantity = new Decimal(minQty);

  let tpsAmount = tpAmounts.map(tp => new Decimal(tp));
  const tpsSum = tpsAmount.reduce((sum, current) => sum.plus(current), new Decimal(0));

  let tpsRatio = tpsAmount.map(tp => tp.div(tpsSum));
  let tpsRounded = tpsRatio.map(ratio => (roundedQuantity.times(ratio)).toFixed(precision, Decimal.ROUND_DOWN));

  // Adjust values if any tpsRounded value is smaller than minQuantity
  tpsRounded = tpsRounded.map(tp => new Decimal(tp).lt(minQuantity) ? minQuantity : new Decimal(tp));

  let tpsSumRounded = tpsRounded.reduce((sum, current) => sum.plus(current), new Decimal(0));

  // Adjust total quantity if necessary
  if (tpsSumRounded.gt(roundedQuantity)) {
    let excess = tpsSumRounded.minus(roundedQuantity);

    for (let i = tpsRounded.length - 1; i >= 0; i--) {
      if (tpsRounded[i].gte(excess)) {
        tpsRounded[i] = tpsRounded[i].minus(excess);
        break;
      } else {
        excess = excess.minus(tpsRounded[i]);
        tpsRounded[i] = new Decimal(0);
      }
    }
  } else if (tpsSumRounded.lt(roundedQuantity)) {
    let deficit = roundedQuantity.minus(tpsSumRounded);

    for (let i = tpsRounded.length - 1; i >= 0; i--) {
      if (tpsRounded[i].gt(0)) {
        tpsRounded[i] = tpsRounded[i].plus(deficit);
        break;
      }
    }
  }

  return tpsRounded.map(val => Number(val.toString()));
}

module.exports = rearrangeTps;
